File: 09-backtracking/00017-Letter_Combinations_Of_A_Phone_Number/00017-letter_combinations_of_a_phone_number.java
// 17: Letter Combinations Of A Phone Number
// https://leetcode.com/problems/letter-combinations-of-a-phone-number


import java.util.List;
import java.util.Map;
import java.util.ArrayList;


class Solution {
    private void backtrack(String digits, int index,  Map<Character, String> m, String current, List<String> result) {
        if (index == digits.length()) {
            result.add(current);
            return;
        }

        String s = m.get(digits.charAt(index));
        for (int i=0; i<s.length(); i++) {
            current += (s.charAt(i));
            backtrack(digits, index + 1, m, current, result);
            current = current.substring(0, current.length() - 1);
        }
    }

    
    // SOLUTION
    public List<String> letterCombinations(String digits) {
        if (digits.length() == 0) {
            return new ArrayList<String>();
        }

        Map<Character, String> m = Map.of(
            '2', "abc",
            '3', "def",
            '4', "ghi",
            '5', "jkl",
            '6', "mno",
            '7', "pqrs",
            '8', "tuv",
            '9', "wxyz"
        );

        List<String> result = new ArrayList<>();
        String current = "";
        backtrack(digits, 0, m, current, result);
        return result;
    }

    public static void main(String[] args) {
        Solution o = new Solution();

        // INPUT
        String digits = "23";

        // OUTPUT
        var result = o.letterCombinations(digits);
        System.out.println(result);
    }
}

File: 09-backtracking/00017-Letter_Combinations_Of_A_Phone_Number/00017-letter_combinations_of_a_phone_number.py
# 17: Letter Combinations Of A Phone Number
# https://leetcode.com/problems/letter-combinations-of-a-phone-number



from typing import List

class Solution:
    # SOLUTION
    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0: 
            return []
        
        m = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "qprs",
            "8": "tuv",
            "9": "wxyz",
        }
        result = []
        current = ""

        def backtrack(index: int, current: str) -> None:
            if len(current) == len(digits):
                result.append(current)
                return
            
            for c in m[digits[index]]:
                backtrack(index+1, current + c)


        backtrack(0, current)
        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    digits = "23"

    # OUTPUT
    result = o.letterCombinations(digits)

    print(result)
